### ISSUE TITLE
ParallelTransferOptions support in download file can improve download file performance?

### ISSUE THREAD

**Issue by harisingh-highq** [Azure/azure-sdk-for-java#19007]
Hi @gapra-msft @rickle-msft 

**Our library version:- azure-storage-blob 12.10.0**

As mentioned here, https://stackoverflow.com/questions/65395726/upload-of-large-files-using-azure-sdk-for-java-with-limited-heap ParallelTransferOptions can work only with blobClient.downloadToFileWithResponse method.

If you see highlighted comment in https://stackoverflow.com/questions/65395726/upload-of-large-files-using-azure-sdk-for-java-with-limited-heap as like below:-
![image](https://user-images.githubusercontent.com/33862788/107024141-a8a0d000-67cd-11eb-91d5-7696642bca52.png)

From above comments, I've below query:-

**After applying ParallelTransferOptions setBlockSize  & setMaxConcurrency, download file performance will be improved in terms of memory and download time?**

We've handled download file in our case in different manner using blobClient.openInputStream() and then read & decrypt stream and write to ServletOutputStream as like below:-
![image](https://user-images.githubusercontent.com/33862788/107023164-41cee700-67cc-11eb-8676-ee4a30306d75.png)

**In above program, how can we set ParallelTransferOptions in download file?**

I found below supported methods only in azure blob storage library for download file:-
![image](https://user-images.githubusercontent.com/33862788/107022344-45ae3980-67cb-11eb-8f5f-0ba4d5752999.png)

Can you please help us into this?
Thanks

**Comment by rickle-msft** [comment]
Hi, @harisingh-highq. Thank you for posting these questions.

> After applying ParallelTransferOptions setBlockSize & setMaxConcurrency, download file performance will be improved in terms of memory and download time?
You'll have to define more clearly what you mean by improved performance. Typically the more memory we're allowed to use, the fast we can go, so you have to decide what the right balance is for you.

> In above program, how can we set ParallelTransferOptions in download file?
As you mentioned, ParallelTransferOptions is only available on downloadToFile. download and openInputStream do not do any work in parallel, so the parallel options don't apply there. If you want to make use of that, you could download in parallel to a file and then open an inputstream from the file.

**Comment by harisingh-highq** [comment]
@rickle-msft 
Thank you for your response.

`You'll have to define more clearly what you mean by improved performance. Typically the more memory we're allowed to use, the fast we can go, so you have to decide what the right balance is for you.`
We mean to ask like if we added ParallelTransferOptions with block size & maxConcurrency in **upload file service**, it manages block accordingly and large file could upload successfully which was not uploaded previously without ParallelTransferOptions.

In large file download also, we face slowness in download file as compared to other storage.
So can we add same ParallelTransferOptions  support for download file performance to make it more faster?

But as you mentioned here `Typically the more memory we're allowed to use, the fast we can go` if we allow more memory in download operations then our application memory may busy to serve files faster and other file operations may get impacted and wait to release memory?

`As you mentioned, ParallelTransferOptions is only available on downloadToFile. download and openInputStream do not do any work in parallel, so the parallel options don't apply there. If you want to make use of that, you could download in parallel to a file and then open an inputstream from the file.`
As I understood from your comment, if we want to support ParallelTransferOptions then first we should download file to a temp file with ParallelTransferOptions and then only we can play with inputstream as per our requirement.

**Comment by rickle-msft** [comment]
@harisingh-highq Increasing the maxConcurrency value on parallelTransferOptions should increase the speed of the download, though it will still depend on the resources of your machine and the size of the file on how much it is affected. In any case, that is the best way to tell the sdk to try to speed things up. 

It is true that allowing the sdk to use more memory may come at the expense of other operations on your machine. This is why we always say we can only tell customers what certain options do and how they may affect behavior, but we cannot offer definitive recommendations. Each environment is too unique, so there will have to be a lot of testing and tweaking involve to optimize.

Your comment about using a temp file is correct.

**Comment by harisingh-highq** [comment]
@rickle-msft 
Thank you for the clarification.

**Comment by rickle-msft** [comment]
Sure thing! I am going to close this issue because I believe your question has been addressed. Please feel free to reach out again if you have further questions.

**Comment by harisingh-highq** [comment]
Hi @rickle-msft 

> We've handled download file in our case in different manner using blobClient.openInputStream() and then read & decrypt stream and write to ServletOutputStream as like below:-

As per attached code in above quote message, we're getting below exception while try to download 1GB/2GB file

![image](https://user-images.githubusercontent.com/33862788/107946297-d4366e00-6fb6-11eb-908f-23c832beb051.png)

Can you please help us to resolve it?
Thanks

**Comment by harisingh-highq** [comment]
@rickle-msft @gapra-msft 
> Your comment about using a temp file is correct.
FYI, This solution is not applicable in our case.

Please give some insights about my last comment.
Thank you

**Comment by rickle-msft** [comment]
@harisingh-highq I have responded to your new issue. We will be diligent in responding to each new issue and comment. Please do not try to have the same conversation on multiple threads as it will make it very difficult to follow and keep the discussion productive.
